File: ios/Classes/IvsPlayerPlugin.swift
import Flutter
import UIKit
import AVKit
import AmazonIVSPlayer

public class IvsPlayerPlugin: NSObject, FlutterPlugin {
//    private static var flutterState:FlutterState!
    
    
    
    public static func register(with registrar: FlutterPluginRegistrar) {
//        flutterState = FlutterState(binaryMessenger: registrar.messenger(), textureRegistry: registrar.textures(),flutterPluginRegistrar: registrar)
        
      
                let factory = FLNativeViewFactory(messenger: registrar.messenger())
                registrar.register(factory, withId: IVSConstants.viewTypeIvsPlayer)
    }
    
    
//
//    func autoQualityMode(mode: AutoQualityModeMessage) throws -> Bool {
//        var v:Bool = true
//        do{
//            v = try (CacheUtil.i.getPlayer(key: Int32(mode.viewId))?.autoQualityMode(mode: mode) ?? false)
//        }catch let error{
//            throw error
//        }
//        return   v
//    }
//
//    func looping(loopingMessage: LoopingMessage) throws -> Bool {
//        var v:Bool = false
//        do{
//            v = try (CacheUtil.i.getPlayer(key: Int32(loopingMessage.viewId))?.looping(loopingMessage: loopingMessage) ?? false)
//        }catch let error{
//            throw error
//        }
//        return   v
//    }
//
//    func mute(mutedMessage: MutedMessage) throws -> Bool {
//        var v:Bool = false
//        do{
//            v = try (CacheUtil.i.getPlayer(key: Int32(mutedMessage.viewId))?.mute(mutedMessage: mutedMessage) ?? false)
//        }catch let error{
//            throw error
//        }
//        return   v
//
//    }
//
//    func playbackRate(playbackRateMessage: PlaybackRateMessage) throws -> Double {
//        var v:Double = 1
//        do{
//            v = try (CacheUtil.i.getPlayer(key: Int32(playbackRateMessage.viewId))?.playbackRate(playbackRateMessage:playbackRateMessage) ?? 1)
//        }catch let error{
//            throw error
//        }
//        return   v
//    }
//
//    func volume(volumeMessage: VolumeMessage) throws -> Double {
//        var v:Double = 1
//        do{
//            v = try (CacheUtil.i.getPlayer(key: Int32(volumeMessage.viewId))?.volume(volumeMessage: volumeMessage) ?? 1)
//        }catch let error{
//            throw error
//        }
//        return   v
//    }
//
//    func videoDuration(viewMessage: ViewMessage) throws -> Double {
//        var v:Double = 1
//        do{
//            v = try (CacheUtil.i.getPlayer(key: Int32(viewMessage.viewId))?.videoDuration(viewMessage:viewMessage) ?? 0)
//        }catch let error{
//            throw error
//        }
//        return   v
//    }
//
//    func playbackPosition(viewMessage: ViewMessage) throws -> Double {
//        var v:Double = 1
//        do{
//            v = try (CacheUtil.i.getPlayer(key: Int32(viewMessage.viewId))?.playbackPosition(viewMessage:viewMessage) ?? 0)
//        }catch let error{
//            throw error
//        }
//        return   v
//    }
//
//    func qualities(viewMessage: ViewMessage) throws -> [FQuality] {
//        var v:[FQuality] = []
//        do{
//            v = try (CacheUtil.i.getPlayer(key: Int32(viewMessage.viewId))?.qualities(viewMessage:viewMessage) ?? [])
//        }catch let error{
//            throw error
//        }
//        return   v
//    }
//
//    func quality(qualityMessage: FQualityMessage) throws -> FQuality {
//        var v:FQuality = FQuality(name: "", height: 0, width: 0)
//        do{
//            v = try (CacheUtil.i.getPlayer(key: Int32(qualityMessage.viewId))?.quality(qualityMessage: qualityMessage) ?? v)
//        }catch let error{
//            throw error
//        }
//        return   v
//    }
//
//    func create() throws -> Int64? {
//
//        return nil
//
//        //    let playerInstance = IvsPlayer( binaryMessenger: flutterState.binaryMessenger)
////      let textureId = flutterState.textureRegistry.register(playerInstance)
////        CacheUtil.i.setPlayer(key: textureId, playerInstance: playerInstance)
//    }
//
//    func pause(viewMessage: ViewMessage) throws {
//        try CacheUtil.i.getPlayer(key: Int32(viewMessage.viewId))?.pause(viewMessage: viewMessage)
//    }
//
//    func load(loadMessage: LoadMessage) throws {
//        try CacheUtil.i.getPlayer(key: Int32(loadMessage.viewId))?.load(loadMessage: loadMessage)
//
//    }
//
//    func play(viewMessage: ViewMessage) throws {
//        try CacheUtil.i.getPlayer(key: Int32(viewMessage.viewId))?.play(viewMessage: viewMessage)
//
//    }
//
//    func seekTo(seekMessage: SeekMessage) throws {
//        try CacheUtil.i.getPlayer(key: Int32(seekMessage.viewId))?.seekTo(seekMessage: seekMessage)
//
//    }
//
//    func dispose(viewMessage: ViewMessage) throws {
//        try CacheUtil.i.getPlayer(key: Int32(viewMessage.viewId))?.dispose(viewMessage: viewMessage)
//        CacheUtil.i.removePlayer(key: Int32(viewMessage.viewId))
//    }
//
//
    
//    private var flutterState:FlutterState!
    
//    public static func register(with registrar: FlutterPluginRegistrar) {
//
//        self.flutterState = FlutterState(binaryMessenger: <#T##FlutterBinaryMessenger#>, textureRegistry: <#T##FlutterTextureRegistry#>)
//
//        registrar.textures();
//        let factory = FLNativeViewFactory(messenger: registrar.messenger())
//        registrar.register(factory, withId: IVSConstants.viewTypeIvsPlayer)
//    }
}


class FLNativeViewFactory: NSObject, FlutterPlatformViewFactory {
    private var messenger: FlutterBinaryMessenger

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
//        return IvsPlayer(
//            frame: frame,
//            viewIdentifier: viewId,
//            arguments: args,
//            binaryMessenger: messenger)
        
//        if(args != nil){
//           return CacheUtil.i.getPlayer(key: (args as? Int64))
//        }
        
        let argu : Dictionary<String, Any>? = args as? Dictionary<String, Any>
        var ivsP:IvsPlayer
        if(argu != nil){
           var viewId = argu!["viewId"] as? Int64
            ivsP =  CacheUtil.i.getPlayer(key:Int32(viewId ?? 0))!
        }else{
            ivsP = IvsPlayer(binaryMessenger: messenger, withFrame: frame, viewIdentifier: viewId, arguments: args)
        }
        
        return ivsP
    }
    
    /// Implementing this method is only necessary when the `arguments` in `createWithFrame` is not `nil`.
    public func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
          return FlutterStandardMessageCodec.sharedInstance()
    }
}
//

//class FlutterState{
//    let binaryMessenger:FlutterBinaryMessenger
//    let textureRegistry:FlutterTextureRegistry
//    let flutterPluginRegistrar:FlutterPluginRegistrar
//
//    init(binaryMessenger: FlutterBinaryMessenger, textureRegistry: FlutterTextureRegistry, flutterPluginRegistrar:FlutterPluginRegistrar) {
//        self.binaryMessenger = binaryMessenger
//        self.textureRegistry = textureRegistry
//        self.flutterPluginRegistrar = flutterPluginRegistrar
//    }
//
//    func startListening(ivsPlayerPlugin:IvsPlayerPlugin,binaryMessenger:FlutterBinaryMessenger){
//        IvsPlayerApiSetup.setUp(binaryMessenger: binaryMessenger, api: ivsPlayerPlugin)
//    }
//
//    func stopListening( binaryMessenger: FlutterBinaryMessenger) {
//        IvsPlayerApiSetup.setUp(binaryMessenger: binaryMessenger,api: nil)
//    }
//}

File: android/src/main/kotlin/com/example/ivs_player/IvsPlayerPlugin.kt
package com.example.ivs_player

import AutoQualityModeMessage
import FQuality
import FQualityMessage
import IvsPlayerApi
import LoadMessage
import LoopingMessage
import MutedMessage
import PlaybackRateMessage
import SeekMessage
import ViewMessage
import VolumeMessage
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.view.TextureRegistry

/** IvsPlayerPlugin */
class IvsPlayerPlugin: FlutterPlugin,IvsPlayerApi {

  private lateinit var flutterState:FlutterState


  private lateinit var streamChannel: EventChannel
  private  var eventSink: EventSink? = null

  override fun onAttachedToEngine(binding: FlutterPluginBinding) {
    flutterState = FlutterState(
      binding.applicationContext,
      binding.binaryMessenger,
      binding.textureRegistry
    )
    flutterState.startListening(this,binding.binaryMessenger)

  }

  override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
    flutterState.stopListening(binding.binaryMessenger);
  }
  override fun autoQualityMode(mode: AutoQualityModeMessage): Boolean {
   return CacheUtil.getPlayer(mode.viewId)?.autoQualityMode(mode) ?: true
  }

  override fun looping(loopingMessage: LoopingMessage): Boolean {
    return CacheUtil.getPlayer(loopingMessage.viewId)?.looping(loopingMessage) ?: true
  }

  override fun mute(mutedMessage: MutedMessage): Boolean {
    return CacheUtil.getPlayer(mutedMessage.viewId)?.mute(mutedMessage) ?: true
  }

  override fun playbackRate(playbackRateMessage: PlaybackRateMessage): Double {
    return CacheUtil.getPlayer(playbackRateMessage.viewId)?.playbackRate(playbackRateMessage) ?: 0.0
  }

  override fun volume(volumeMessage: VolumeMessage): Double {
    return CacheUtil.getPlayer(volumeMessage.viewId)?.volume(volumeMessage) ?: 0.0
  }

  override fun videoDuration(viewMessage: ViewMessage): Double {
    return CacheUtil.getPlayer(viewMessage.viewId)?.videoDuration(viewMessage) ?: 0.0
  }

  override fun playbackPosition(viewMessage: ViewMessage): Double {
    return CacheUtil.getPlayer(viewMessage.viewId)?.playbackPosition(viewMessage) ?: 0.0
  }

  override fun qualities(viewMessage: ViewMessage): List<FQuality> {
    return CacheUtil.getPlayer(viewMessage.viewId)?.qualities(viewMessage) ?: emptyList()
  }

  override fun quality(qualityMessage: FQualityMessage): FQuality {
    return CacheUtil.getPlayer(qualityMessage.viewId)?.quality(qualityMessage) ?: FQuality(  "",0, 0)
  }

  override fun create(): Long {
    val texture = flutterState.textureRegistry.createSurfaceTexture();
    CacheUtil.setPlayer(texture.id() ?: 0,IvsPlayer(flutterState.context,texture,flutterState.binaryMessenger));
    return  texture.id()
  }


  override fun pause(viewMessage: ViewMessage) {
    CacheUtil.getPlayer(viewMessage.viewId)?.pause(viewMessage)
  }

  override fun load(loadMessage: LoadMessage) {
    CacheUtil.getPlayer(loadMessage.viewId)?.load(loadMessage)
  }

  override fun play(viewMessage: ViewMessage) {
    CacheUtil.getPlayer(viewMessage.viewId)?.play(viewMessage)
  }

  override fun seekTo(seekMessage: SeekMessage) {
    CacheUtil.getPlayer(seekMessage.viewId)?.seekTo(seekMessage)
  }

  override fun dispose(viewMessage: ViewMessage) {
    CacheUtil.getPlayer(viewMessage.viewId)?.dispose()
    CacheUtil.removePlayer(viewMessage.viewId)
  }
}

class FlutterState(val context: Context,val binaryMessenger: BinaryMessenger,val textureRegistry: TextureRegistry){
  fun startListening(ivsPlayerPlugin: IvsPlayerPlugin, binaryMessenger: BinaryMessenger): Unit {
    IvsPlayerApi.setUp(binaryMessenger,ivsPlayerPlugin)
  }

  fun stopListening( binaryMessenger: BinaryMessenger): Unit {
    IvsPlayerApi.setUp(binaryMessenger,null)
  }
}